### PR TITLE
fix(amplify-category-auth): update front end config on pull

### PR DESCRIPTION
**Pull request by lazpavel**
#### Description of changes
- ensure that for projects created before 5.2.0, on pull we populate the auth frontend config 

#### Description of how you validated changes
- `yarn test` passes,
- manual testing by creating a project with an older Amplify CLI and pulling into an empty folder populates the front end configuration

#### Checklist
- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ericclemmons** [comment]
When running `$(yarn global bin)/amplify-dev pull` against my backend created via Amplify CLI in July (https://github.com/aws-amplify/amplify-ui/tree/main/environments/auth-with-username-no-attributes), I'm getting the following `aws-exports.js` **with an empty `aws_cognito_login_mechanisms`**:

```js
/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "us-east-1",
    "aws_cognito_identity_pool_id": "us-east-1:******",
    "aws_cognito_region": "us-east-1",
    "aws_user_pools_id": "us-east-1_******",
    "aws_user_pools_web_client_id": "******",
    "oauth": {},
    "aws_cognito_login_mechanisms": [],
    "aws_cognito_signup_attributes": [
        "EMAIL"
    ],
    "aws_cognito_mfa_configuration": "OFF",
    "aws_cognito_mfa_types": [
        "SMS"
    ],
    "aws_cognito_password_protection_settings": {
        "passwordPolicyMinLength": 8,
        "passwordPolicyCharacters": []
    }
};


export default awsmobile;

```

Is `username` a default value somewhere?  It shows up correctly in https://us-east-1.admin.amplifyapp.com/admin/d2zuq5rjkps8u4/dev/auth:

> Screen Shot 2021-09-03 at 10 36 46 AM

This what's in my `parameters.json`:

```json
{
    "identityPoolName": "authwithusernamenoat3aaa0954_identitypool_3aaa0954",
    "allowUnauthenticatedIdentities": false,
    "resourceNameTruncated": "authwi3aaa0954",
    "userPoolName": "authwithusernamenoat3aaa0954_userpool_3aaa0954",
    "autoVerifiedAttributes": [
        "email"
    ],
    "mfaConfiguration": "OFF",
    "mfaTypes": [
        "SMS Text Message"
    ],
    "smsAuthenticationMessage": "Your authentication code is {####}",
    "smsVerificationMessage": "Your verification code is {####}",
    "emailVerificationSubject": "Your verification code",
    "emailVerificationMessage": "Your verification code is {####}",
    "defaultPasswordPolicy": false,
    "passwordPolicyMinLength": 8,
    "passwordPolicyCharacters": [],
    "requiredAttributes": [
        "email"
    ],
    "userpoolClientGenerateSecret": false,
    "userpoolClientRefreshTokenValidity": 30,
    "userpoolClientWriteAttributes": [
        "email"
    ],
    "userpoolClientReadAttributes": [
        "email"
    ],
    "userpoolClientLambdaRole": "authwi3aaa0954_userpoolclient_lambda_role",
    "userpoolClientSetAttributes": false,
    "sharedId": "3aaa0954",
    "resourceName": "authwithusernamenoat3aaa0954",
    "authSelections": "identityPoolAndUserPool",
    "authRoleArn": {
        "Fn::GetAtt": [
            "AuthRole",
            "Arn"
        ]
    },
    "unauthRoleArn": {
        "Fn::GetAtt": [
            "UnauthRole",
            "Arn"
        ]
    },
    "useDefault": "default",
    "userPoolGroupList": [],
    "serviceName": "Cognito",
    "usernameCaseSensitive": false,
    "dependsOn": []
}
```

I don't see a `username` specified anywhere in `paramters.json` nor my cloudformation template.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 1]
```diff
@@ -106,10 +106,24 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
...
+      usernameAttributes
+        .toUpperCase()
+        .replace(/\s/g, '')
+        .split(',')
+        .forEach(it => loginMechanisms.push(it));
+    } else {
+      loginMechanisms.push(usernameAttributes.toUpperCase());
+    }
+  }
+
```
Comment (ericclemmons) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:126]: Do we need to do a check for `if (loginMechanisms.length === 0)` and add the username as a default?

See: https://github.com/aws-amplify/amplify-cli/pull/8093#issuecomment-912630619

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 2]
```diff
@@ -11,54 +11,54 @@ import { AuthParameters } from '../import/types';
...
-  options.customAuth = customAuthConfigured;
-  options.frontendAuthConfig = getFrontendConfig(authParameters);
+export const getPostAddAuthMetaUpdater =
+  (context: any, resultMetadata: { service: string; providerName: string }) =>
+  (resourceName: string): string => {
+    const options: any = {
+      service: resultMetadata.service,
+      providerPlugin: resultMetadata.providerName,
+    };
+    const parametersJSONPath = path.join(context.amplify.pathManager.getBackendDirPath(), 'auth', resourceName, 'parameters.json');
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:21]: We have a method for this in `stateManager` to read `parameters.json`

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 3]
```diff
@@ -11,54 +11,54 @@ import { AuthParameters } from '../import/types';
...
+            attributes: ['UserPoolId', 'AppClientIDWeb', 'AppClientID'],
+          },
+        ];
+        context.amplify.updateamplifyMetaAfterResourceUpdate('auth', 'userPoolGroups', 'dependsOn', userPoolGroupDependsOn);
+      }
     }
-  }
-  return resourceName;
-};
+    return resourceName;
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:60]: GH shows difference here, but I cannot spot any, could you revert this section if there are no changes? Lines 22-60

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 4]
```diff
@@ -106,10 +106,24 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
...
 };
 
-function getFrontendConfig(authParameters: AuthParameters) {
+export function getFrontendConfig(authParameters: AuthParameters) {
   const loginMechanisms: string[] = [];
   loginMechanisms.push(...(authParameters?.aliasAttributes || []).map((att: string) => att.toUpperCase()));
 
+  // backwards compatibility
+  if (authParameters?.usernameAttributes) {
+    const usernameAttributes = authParameters?.usernameAttributes[0];
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:115]: can we just do a plain split and process the result and remove regex magic? that would be a much simpler code path IMHO, does this code guarantees to NOT to end up duplicate entries somehow? Perhaps using a `Set` instead of pure array would protect against that

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts [hunk 5]
```diff
@@ -63,3 +66,25 @@ function attachContextExtensions(context, packageLocation) {
     }
   }
 }
+
+function migrateAmplifyMeta(context, amplifyMeta?) {
```
Comment (attilah) [inline packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts:70]: We need a more specific name here and some comments on what it migrated FROM and TO, feel free to link issue # as well.

This only invoked when a category is changing, what happens to customers who are not running a `push` or `update auth`, just doing a `pull` over the existing, already initialized project with the new CLI that has these changes?

If we have a migration of state files in a PR we have to add migration tests to cover and verify that everything works correctly.

File: yarn.lock [hunk 6]
```diff
@@ -16745,6 +16745,13 @@ node-pty-prebuilt-multiarch@^0.9.0:
     nan "^2.14.0"
     prebuild-install "^5.2.5"
 
+node-pty@^0.10.0:
```
Comment (attilah) [inline yarn.lock:16748]: these changes are not related to the PR, remove them

**Review by attilah — CHANGES_REQUESTED** [review]
@ericclemmons asked me to take a look at the PR so even if it is a draft I reviewed it to give some directions for the future changes.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts (hunk 3 above)
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:60]: this changes seems to be triggered by a commit formatting hook, will update

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts (hunk 4 above)
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts:115]: updated

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts [hunk 7]
```diff
@@ -63,3 +66,32 @@ function attachContextExtensions(context, packageLocation) {
...
+    });
+
+    if (!authResourceName) return;
+
+    const authParameters: AuthParameters = stateManager.getResourceParametersJson(undefined, 'auth', authResourceName);
+    const frontendAuthConfig = getFrontendConfig(authParameters);
+
+    amplifyMeta.auth[authResourceName].frontendAuthConfig = amplifyMeta.auth[authResourceName].frontendAuthConfig ?? {};
+    Object.keys(frontendAuthConfig).forEach(function (key) {
+      if (!amplifyMeta.auth[authResourceName].frontendAuthConfig.hasOwnProperty(key)) {
```
Comment (cjihrig) [inline packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts:90]: You should be able to cache `amplifyMeta.auth[authResourceName].frontendAuthConfig` so you don't have to keep computing it in a loop.

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts [hunk 8]
```diff
@@ -63,3 +66,32 @@ function attachContextExtensions(context, packageLocation) {
...
+      return amplifyMeta.auth[key].service === 'Cognito';
+    });
+
+    if (!authResourceName) return;
+
+    const authParameters: AuthParameters = stateManager.getResourceParametersJson(undefined, 'auth', authResourceName);
+    const frontendAuthConfig = getFrontendConfig(authParameters);
+
+    amplifyMeta.auth[authResourceName].frontendAuthConfig = amplifyMeta.auth[authResourceName].frontendAuthConfig ?? {};
+    Object.keys(frontendAuthConfig).forEach(function (key) {
```
Comment (cjihrig) [inline packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts:89]: Please make this an arrow function.

**Review by cjihrig — COMMENTED** [review]
Is it possible to add a test for this?

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts [hunk 9]
```diff
@@ -2,7 +2,9 @@ import * as fs from 'fs-extra';
 import * as path from 'path';
 import { getResourceOutputs } from './get-resource-outputs';
 import sequential from 'promise-sequential';
-import { stateManager } from 'amplify-cli-core';
+import { JSONUtilities, stateManager } from 'amplify-cli-core';
+import { AuthParameters } from 'amplify-category-auth/lib/provider-utils/awscloudformation/import/types';
+import { getFrontendConfig } from 'amplify-category-auth/lib/provider-utils/awscloudformation/utils/amplify-meta-updaters';
```
[the remark this answers is not in the document]
Comment (attilah) [reply packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts:7]: No import should be done from `lib` ever, I agree, VSCode does this perhaps we miss some vscode specific config for our packages to exclude lib for autocompletion and such things.

**Review by attilah — APPROVED** [review]
I agree with @cjihrig's comments, please fix those before merging.

Regarding the awsexports (and others) generation make sure we have a test that preserves the custom values if they were present and also have a test for removal, when someone removes auth and such, then the configured values MUST be removed from awsexports (and others)

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 10]
```diff
@@ -106,20 +106,34 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
   return resourceName;
 };
 
-function getFrontendConfig(authParameters: AuthParameters) {
-  const loginMechanisms: string[] = [];
-  loginMechanisms.push(...(authParameters?.aliasAttributes || []).map((att: string) => att.toUpperCase()));
+export function getFrontendConfig(authParameters: AuthParameters) {
+  const loginMechanisms: Set<string> = new Set<string>();
+  (authParameters?.aliasAttributes || []).forEach(it => loginMechanisms.add(it.toUpperCase()));
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: updated

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 11]
```diff
@@ -106,20 +106,34 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
...
+
+  // backwards compatibility
+  if (authParameters?.usernameAttributes && authParameters?.usernameAttributes.length) {
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: updated

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 12]
```diff
@@ -106,20 +106,34 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
...
+    const usernameAttributes = authParameters?.usernameAttributes[0];
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: updated

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 13]
```diff
@@ -106,20 +106,34 @@ export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceNam
...
+    if (usernameAttributes.includes(',')) {
+      usernameAttributes.split(',').forEach(it => loginMechanisms.add(it.trim().toUpperCase()));
+    } else {
+      loginMechanisms.add(usernameAttributes.toUpperCase());
+    }
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: updated

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts]: thank you, updated

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts [hunk 14]
```diff
@@ -63,3 +66,32 @@ function attachContextExtensions(context, packageLocation) {
...
+
+// projects created before 5.2.0 didn't populate frontend config in amplify-meta.json
+// this method ensures frontend config settings are added to amplify meta on pull as they exist in parameters.json
+// https://app.asana.com/0/1200585422384147/1200740448709567/f
+function ensureAmplifyMetaFrontendConfig(context, amplifyMeta?) {
+  if (!amplifyMeta) {
+    amplifyMeta = stateManager.getMeta();
+  }
+
+  if (amplifyMeta.auth) {
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts]: updated

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts (hunk 7 above)
Comment (lazpavel) [reply packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts]: updated

File: packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts (hunk 8 above)
Comment (lazpavel) [reply packages/amplify-cli/src/extensions/amplify-helpers/on-category-outputs-change.ts]: updated

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts [hunk 15]
```diff
@@ -139,7 +153,7 @@ function getFrontendConfig(authParameters: AuthParameters) {
   }
 
   return {
-    loginMechanisms: loginMechanisms,
+    loginMechanisms: Array.from(loginMechanisms.values()),
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: updated

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/amplify-meta-updaters.ts]: switched to `if (authParameters?.usernameAttributes && authParameters?.usernameAttributes.length > 0) {` to avoid truthy evaluation
